src/mame/namco/namcos10_exio.cpp

```diff
@@ -0,0 +1,210 @@
+
+	if (data == 3) {
+		// Probably this should just reset the entire CPU once the program is loaded in memory
+		// The CPU was halted as soon as it was started so just set the PC and resume it.
+		auto pc = ((m_ram[0x4f02 / 2] << 16) | m_ram[0x4f00 / 2]) & 0xffffff;
+		m_maincpu->set_state_int(TLCS900_PC, pc);
+		m_maincpu->resume(SUSPEND_REASON_HALT);
+	}
+}
+
+void namcos10_exio_device::ram_w(offs_t offset, uint16_t data, uint16_t mem_mask)
+{
+	if (BIT(offset, 0))
+		m_ram[offset / 2] = ((data & 0xff) << 8) | (m_ram[offset / 2] & 0xff);
+	else
+		m_ram[offset / 2] = (m_ram[offset / 2] & 0xff00) | (data & 0xff);
+}
+
+uint16_t namcos10_exio_device::ram_r(offs_t offset)
+{
+	if (BIT(offset, 0))
+		return BIT(m_ram[offset / 2], 8, 8);
+
+	return BIT(m_ram[offset / 2], 0, 8);
+}
+
+template <int port>
+void namcos10_exio_device::port_write(offs_t offset, uint8_t data)
+{
+	logerror("%s: exio_port%d_write %02x\n", machine().describe_context(), port, data);
+
+	if (port == 8) {
+		// HACK: Simple check to just know when the CPU is alive
+		m_is_active |= data != 0;
+	}
+}
+
+template <int port>
+uint8_t namcos10_exio_device::port_read(offs_t offset)
+{
+	logerror("%s: exio_port%d_read\n", machine().describe_context(), port);
+	return 0;
+}
+
+////////////////////////////////////
+
+namcos10_mgexio_device::namcos10_mgexio_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+	: namcos10_exio_base_device(mconfig, NAMCOS10_MGEXIO, tag, owner, clock, 0x33),
+	m_maincpu(*this, "exio_mcu"),
+	m_ram(*this, "exio_ram"),
+	m_nvram(*this, "nvram")
+{
+}
+
+void namcos10_mgexio_device::device_start()
+{
+	save_item(NAME(m_active_state));
+	m_nvram->set_base(m_ram, 0x8000);
+}
+
+void namcos10_mgexio_device::device_reset()
+{
+}
+
+void namcos10_mgexio_device::device_reset_after_children()
+{
+	m_maincpu->suspend(SUSPEND_REASON_HALT, 1);
+	m_active_state = 0;
+}
+
+void namcos10_mgexio_device::map(address_map &map)
+{
+	map(0x00000, 0x7ffff).ram().share(m_ram);
+}
+
+void namcos10_mgexio_device::io_map(address_map &map)
+{
+	// TODO: Hook up I/O ports
+}
+
+void namcos10_mgexio_device::device_add_mconfig(machine_config &config)
+{
+	H83007(config, m_maincpu, 14.746_MHz_XTAL);
+	m_maincpu->set_mode_a20();
+	m_maincpu->set_addrmap(AS_PROGRAM, &namcos10_mgexio_device::map);
+	m_maincpu->set_addrmap(AS_IO, &namcos10_mgexio_device::io_map);
+
+	NVRAM(config, m_nvram, nvram_device::DEFAULT_ALL_0);
+}
+
+uint16_t namcos10_mgexio_device::cpu_status_r()
+{
+	// pacmball's code call bit 1 the "sub_cpu_enable_flag"
+	auto r = m_active_state == 2 ? 2 : 0;
+
+	if (m_active_state == 1)
+		m_active_state = 2; // needs to see this as not active at least once
+
+	return r;
+}
+
+void namcos10_mgexio_device::ctrl_w(uint16_t data)
+{
+	logerror("%s: exio_ctrl_w %04x\n", machine().describe_context(), data);
+
+	if (data == 3) {
+		m_maincpu->resume(SUSPEND_REASON_HALT);
+		m_active_state = 1;
+	}
```

Comment:
Could you use a timer to activate the CPU and avoid the read-counting hack?  Presumably in real life it takes some time for the clock to stabilise and the CPU to reset.

Reply:
I can play around with a timer perhaps. The game I was working on had a set up that it could check to make sure it wasn't active and then wait for it to be active. If it somehow became active before it started checking the status register it'd just fail entirely so I think there's probably some expected delay there.

Reply:
I'll come back to this soon-ish (sometime within the week), wait for this before merging. This is my first time working with H8 anything so I think I just missed a lot. Going to work through the MGEXIO's init routine and see if there's something more obvious that should be the trigger.